provider/oci/common/client.go

```diff
@@ -0,0 +1,198 @@
+
+package common
+
+import (
+	"context"
+	"crypto/rsa"
+	"fmt"
+
+	"github.com/juju/errors"
+	"github.com/juju/juju/instance"
+	"github.com/juju/juju/network"
+
+	ociCommon "github.com/oracle/oci-go-sdk/common"
+	ociCore "github.com/oracle/oci-go-sdk/core"
+	ociIdentity "github.com/oracle/oci-go-sdk/identity"
+)
+
+type jujuConfigProvider struct {
+	key            []byte
+	keyFingerprint string
+	passphrase     string
+	tenancyOCID    string
+	userOCID       string
+	region         string
+}
+
+type ociClient struct {
+	ociCore.ComputeClient
+	ociCore.BlockstorageClient
+	ociCore.VirtualNetworkClient
+	ociIdentity.IdentityClient
+
+	ociCommon.ConfigurationProvider
+}
+
+// NewJujuConfigProvider returns a new ociCommon.ConfigurationProvider instance
+func NewJujuConfigProvider(user, tenant string, key []byte, fingerprint, passphrase, region string) ociCommon.ConfigurationProvider {
+	return &jujuConfigProvider{
+		key:            key,
+		keyFingerprint: fingerprint,
+		passphrase:     passphrase,
+		tenancyOCID:    tenant,
+		userOCID:       user,
+		region:         region,
+	}
+}
+
+func NewOciClient(provider ociCommon.ConfigurationProvider) (ApiClient, error) {
+	computeClient, err := ociCore.NewComputeClientWithConfigurationProvider(provider)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	blockStorage, err := ociCore.NewBlockstorageClientWithConfigurationProvider(provider)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	virtualNetwork, err := ociCore.NewVirtualNetworkClientWithConfigurationProvider(provider)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	ident, err := ociIdentity.NewIdentityClientWithConfigurationProvider(provider)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	return &ociClient{
+		computeClient,
+		blockStorage,
+		virtualNetwork,
+		ident,
+		provider,
+	}, nil
+}
+
+// Ping validates that the client can access the OCI API successfully
+func (o *ociClient) Ping() error {
+	tenancyID, err := o.TenancyOCID()
+	if err != nil {
+		return errors.Trace(err)
+	}
+	request := ociIdentity.ListCompartmentsRequest{
+		CompartmentId: &tenancyID,
+	}
+	ctx := context.Background()
+	_, err = o.ListCompartments(ctx, request)
+	return err
+}
+
+func (o *ociClient) GetInstanceVnicAttachments(instanceID instance.Id, compartmentID *string) (ociCore.ListVnicAttachmentsResponse, error) {
+	instID := string(instanceID)
+	request := ociCore.ListVnicAttachmentsRequest{
+		CompartmentId: compartmentID,
+		InstanceId:    &instID,
+	}
+	response, err := o.ListVnicAttachments(context.Background(), request)
+	if err != nil {
+		return ociCore.ListVnicAttachmentsResponse{}, errors.Trace(err)
+	}
+	return response, nil
+}
+
+func (o *ociClient) GetInstanceVnics(vnics []ociCore.VnicAttachment) ([]ociCore.GetVnicResponse, error) {
+	result := []ociCore.GetVnicResponse{}
+
+	for _, val := range vnics {
+		vnicID := val.VnicId
+		request := ociCore.GetVnicRequest{
```

Comment:
I'm presuming that none of their APIs allow you to pass in multiple parameters to query in one go?
Or would there be a way that we could send a list of vnicID to the API ?

Or even better if we could ask about an instance and have it expand all of the attributes of that instance that we care about.

Reply:
not really. We can "show" details for one instance, or list all instances. We can't even filter instances based on tags via API calls yet. We do filtering client side.

All instance details that can be listed are detailed here:

https://docs.us-phoenix-1.oraclecloud.com/api/#/en/iaas/20160918/Instance/

VNIC details are exposed through a different API call:

https://docs.us-phoenix-1.oraclecloud.com/api/#/en/iaas/20160918/Vnic/GetVnic